### ISSUE TITLE
Updated pdf fields don't show up when page is written

### ISSUE THREAD

**Issue by segevmalool** [py-pdf/pypdf#355]
I'd like to use PyPDF2 to fill out a pdf form. So far, everything is going smoothly, including updating the field text. But when I write the pdf to a file, there is apparently no change in the form. Running this code:

```python
import datetime as dt
from PyPDF2 import PdfFileReader, PdfFileWriter
import re

form701 = PdfFileReader('ABC701LG.pdf')
page = form701.getPage(0)
filled = PdfFileWriter()

#removing extraneous fields
r = re.compile('^[0-9]')
fields = sorted(list(filter(r.match, form701.getFields().keys())), key = lambda x: int(x[0:2]))

filled.addPage(page)
filled.updatePageFormFieldValues(filled.getPage(0), 
                                 {fields[0]: 'some filled in text'})

print(filled.getPage(0)['/Annots'][0].getObject()['/T'])
print(filled.getPage(0)['/Annots'][0].getObject()['/V'])

with open('test.pdf','wb') as fp:
    filled.write(fp)
```
prints text: 

> 1  EFFECTIVE DATE OF THIS SCHEDULE <i.e. the field name>
> some filled in text

But when I open up test.pdf, there is no added text on the page! Help!

**Comment by mwhit74** [comment]
I am having this same issue. The data does not show up in Adobe Reader unless you activate the field. The data does show up in Bluebeam but if you print, flatten, or push the pdf to a studio session all the data is lost. 

When the file is opened in Bluebeam it automatically thinks that the user has made changes, denoted by the asterisk next to the file name in the tab.

If you export the fdf file from Bluebeam all the data is in the fdf file in the proper place.

If you change any attribute of the field in Bluebeam or Adobe, it will recognize the text in that field. It will print correctly and flatten correctly. I am not sure if it will push to the Bluebeam studio but I assume it will. You can also just copy and paste the text in the field back into that field and it will render correctly. 

I have not found any help after googling around all day. I think it is an issue with PyPDF2 not "redrawing" the PDF correctly. 

I have contacted Bluebeam support and they have returned saying essentially that it is not on their end.

**Comment by mwhit74** [comment]
Ok I think I have narrowed this down some by just comparing two different pdfs.

For reference I am trying to read a pdf that was originally created by Bluebeam, use the updatePageFormFields() function in PyPDF2 to push a bunch of data from a database into the form fields, and save. At some point we want to flatten these and that is when it all goes wrong in Bluebeam. In Adobe it is messed up from the start in that you don't see any values in the form fields until you scroll over them with the mouse. 

I appears there is a problem with the stream object that follows the object(s) representing the text form field. See below. 

This is a sample output from a pdf generated by PyPDF2 for a text form field:
```
26 0 obj<</Subtype/Widget/M(D:20160512102729-05'00')/NM(OEGVASQHFKGZPSZW)/MK<</IF<</A[0 0]>>>>/F 4/C[1 0 0]/Rect[227.157 346.3074 438.2147 380.0766]/V(Marshall CYG)/Type/Annot/FT/Tx/AP<</N 27 0 R>>/DA(0 0 0 rg /Helv 12 Tf)/T(Owner Group)/BS 29 0 R/Q 0/P 3 0 R>>
endobj
27 0 obj<</Type/XObject/Matrix[1 0 0 1 0 0]/Resources<</ProcSet[/PDF/Text]/Font<</Helv 28 0 R>>>>/Length 41/FormType 1/BBox[0 0 211.0577 33.76923]/Subtype/Form>>
stream
0 0 211.0577 33.76923 re W n /Tx BMC EMC 
endstream
endobj
28 0 
'''

And if I back up and edit the same based file in Bluebeam the output from that pdf for a text form field looks like this (I think the border object can be ignored):
```
16 0 obj<</Type/Annot/P 5 0 R/F 4/C[1 0 0]/Subtype/Widget/Q 0/FT/Tx/T(Owner Group)/MK<</IF<</A[0 0]>>>>/DA(0 0 0 rg /Helv 12 Tf)/AP<</N 18 0 R>>/M(D:20170906125217-05'00')/Rect[227.157 346.3074 438.2147 380.0766]/NM(OEGVASQHFKGZPSZW)/BS 17 0 R/V(Marshall CYG)>>
endobj
17 0 obj<</W 1/S/S/Type/Border>>
endobj
18 0 obj<</Type/XObject/Subtype/Form/FormType 1/BBox[0 0 211.0577 33.7692]/Resources<</ProcSet[/PDF/Text]/Font<</Helv 12 0 R>>>>/Matrix[1 0 0 1 0 0]/Length 106>>
stream
0 0 211.0577 33.7692 re W n /Tx BMC BT 0 0 0 rg /Helv 12 Tf 1 0 0 1 2 12.6486 Tm (Marshall CYG) Tj ET EMC 
endstream
```

Ok so the biggest difference here is the stream object at the end. The value /V(Marshall CYG) gets updated in the first object of each pdf, objects 26 and 16 respectively. However the stream object in the PyPDF2 generated pdf does not get updated and the stream object from Bluebeam does get updated. 

In testing this theory I made a copy of the PyPDF2 pdf and manually edited the stream object in a text editor. I open this new file in Bluebeam and flattened it. It worked. This also appears to work in adobe reader. 

Now how to fix....

**Comment by ademidun** [comment]
A potential solution seems to be setting the Need Appearances flag.
Not yet sure how to implement in pypdf2 but these 2 links may provide some clues:
https://stackoverflow.com/questions/12198742/pdf-form-text-hidden-unless-clicked
https://forums.adobe.com/thread/305250

**Comment by ademidun** [comment]
Okay, I think I have figured it out. If you read section 12.7.2 of the [PDF 1.7 specification](adobe.com/content/dam/Adobe/en/devnet/acrobat/pdfs/PDF32000_2008.pdf), you will see that you need to set the NeedAppearances flag of the Acroform.

```
pdf = PdfFileReader(open(infile, 'rb'), strict=False)

if "/AcroForm" in pdf.trailer["/Root"]:
    pdf.trailer["/Root"]["/AcroForm"].update(
        {NameObject("/NeedAppearances"): BooleanObject(True)}
    )
pdf = PdfFileWriter()

if "/AcroForm" in pdf._root_object:
    pdf._root_object["/AcroForm"].update(
        {NameObject("/NeedAppearances"): BooleanObject(True)}
    )
```

**Comment by Tromar44** [comment]
ademidun - Can you elaborate on your suggested solution above? I too am having problems with pdf forms, edited with PyPDF2, not showing field values without clicking in the field. With the code example below, how do you "set the NeedAppearances flag of the Acroform"?
```
from PyPDF2 import PdfFileWriter, PdfFileReader

output = PdfFileWriter()
input = PdfFileReader(open("myInputPdf.pdf", "rb"))

field_dictionary = {'Make': 'Toyota', 'Model': 'Tacoma'}

for pageNum in range(input.numPages):
    pageObj = input.getPage(pageNum)
    output.addPage(pageObj)
    output.updatePageFormFieldValues(pageObj, field_dictionary)

outputStream = open("myOutputPdf.pdf", "wb")
output.write(outputStream)
```
Thanks for any help!

**Comment by ademidun** [comment]
@Tromar44  Preamble, make sure your form is interactive. E.g. The pdf must already have editable fields.

1) Sorry forgot to mention you will have to import them:
`from PyPDF2.generic import BooleanObject, NameObject, IndirectObject`
2) Are you sure you are using `output.__root_object["/AcroForm"]` or `output.trailer["/Root"]["/AcroForm"]` to access the "/AcroForm" key? and not just doing `output["/AcroForm"]`

**Comment by Tromar44** [comment]
@ademidun I thank you very much for your help but unfortunately I'm still not having any luck. To be clear, my simple test pdf form does have two editable fields and the script will populate them with "Toyota" and "Tacoma" respectively but those values are not visible unless I click on the field in the form (they become invisible again after the field loses focus). Here is the rewritten code that includes your suggestions and the results of running the code in inline comments.
```
from PyPDF2 import PdfFileWriter, PdfFileReader
from PyPDF2.generic import BooleanObject, NameObject, IndirectObject

infile = "myInputPdf.pdf"
outfile = "myOutputPdf.pdf"

pdf = PdfFileReader(open(infile, "rb"), strict=False)
if "/AcroForm" in pdf.trailer["/Root"]: # result: following "IF code is executed
    print(True)
    pdf.trailer["/Root"]["/AcroForm"].update(
        {NameObject("/NeedAppearances"): BooleanObject(True)})

pdf2 = PdfFileWriter()
if "/AcroForm" in pdf2._root_object: # result: False - following "IF" code is NOT executed
    print(True)
    pdf2._root_object["/AcroForm"].update(
        {NameObject("/NeedAppearances"): BooleanObject(True)})

if "/AcroForm" in pdf2._root_object["/AcroForm"]: # result: "KeyError: '/AcroForm'
    print(True)
    pdf2._root_object["/AcroForm"].update(
        {NameObject("/NeedAppearances"): BooleanObject(True)})

if "/AcroForm" in pdf2.trailer["/Root"]["/AcroForm"]:  # result: AttributeError: 'PdfFileWriter' object has no attribute 'trailer'
    print(True)
    pdf2._root_object["/AcroForm"].update(
        {NameObject("/NeedAppearances"): BooleanObject(True)})

field_dictionary = {"Make": "Toyota", "Model": "Tacoma"}

pdf2.addPage(pdf.getPage(0))
pdf2.updatePageFormFieldValues(pdf2.getPage(0), field_dictionary)

outputStream = open(outfile, "wb")
pdf2.write(outputStream)
```
I would definitely appreciate any more suggestions that you may have! Thank you very much!

**Comment by ademidun** [comment]
It may also be a browser issue. I don't have the links anymore but I remember reading about some issues when opening/creating a PDF on Preview on Mac or viewing it in the browser vs. using an Adobe app etc. Maybe if you google things like "form fields only showing on click" or "form fields only active on click using preview mac".

 I also recommend reading the [PDF spec link](http://www.adobe.com/content/dam/acom/en/devnet/acrobat/pdfs/PDF32000_2008.pdf) I posted, its a bit dense but a combination of all these should get you in the right direction.

**Comment by ademidun** [comment]
@Tromar44 Okay, I also found this snippet from my code, maybe it will help:

```
def set_need_appearances_writer(writer: PdfFileWriter):
    # See 12.7.2 and 7.7.2 for more information: http://www.adobe.com/content/dam/acom/en/devnet/acrobat/pdfs/PDF32000_2008.pdf
    try:
        catalog = writer._root_object
        # get the AcroForm tree
        if "/AcroForm" not in catalog:
            writer._root_object.update({
                NameObject("/AcroForm"): IndirectObject(len(writer._objects), 0, writer)
            })

        need_appearances = NameObject("/NeedAppearances")
        writer._root_object["/AcroForm"][need_appearances] = BooleanObject(True)
        # del writer._root_object["/AcroForm"]['NeedAppearances']
        return writer

    except Exception as e:
        print('set_need_appearances_writer() catch : ', repr(e))
        return writer
```

**Comment by Tromar44** [comment]
@ademidun That worked perfectly (I'd high five you right now if I could)! Thank you very much! For anyone else interested, the following worked for me:
```
from PyPDF2 import PdfFileWriter, PdfFileReader
from PyPDF2.generic import BooleanObject, NameObject, IndirectObject

def set_need_appearances_writer(writer: PdfFileWriter):
    # See 12.7.2 and 7.7.2 for more information: http://www.adobe.com/content/dam/acom/en/devnet/acrobat/pdfs/PDF32000_2008.pdf
    try:
        catalog = writer._root_object
        # get the AcroForm tree
        if "/AcroForm" not in catalog:
            writer._root_object.update({
                NameObject("/AcroForm"): IndirectObject(len(writer._objects), 0, writer)})

        need_appearances = NameObject("/NeedAppearances")
        writer._root_object["/AcroForm"][need_appearances] = BooleanObject(True)
        return writer

    except Exception as e:
        print('set_need_appearances_writer() catch : ', repr(e))
        return writer

infile = "myInputPdf.pdf"
outfile = "myOutputPdf.pdf"

pdf = PdfFileReader(open(infile, "rb"), strict=False)
if "/AcroForm" in pdf.trailer["/Root"]:
    pdf.trailer["/Root"]["/AcroForm"].update(
        {NameObject("/NeedAppearances"): BooleanObject(True)})

pdf2 = PdfFileWriter()
if "/AcroForm" in pdf2._root_object:
    pdf2._root_object["/AcroForm"].update(
        {NameObject("/NeedAppearances"): BooleanObject(True)})

field_dictionary = {"Make": "Toyota", "Model": "Tacoma"}

pdf2.addPage(pdf.getPage(0))
pdf2.updatePageFormFieldValues(pdf2.getPage(0), field_dictionary)

outputStream = open(outfile, "wb")
pdf2.write(outputStream)
```

**Comment by kissmett** [comment]
@ademidun you great!!!

**Comment by caver456** [comment]
Just stumbled upon this solution - great work!  A couple of issues I noticed - can you reproduce them? - won't have time to send test case details for a couple of days yet if you need them; we had been using the good-ol fdfgen-then-pdftk-subprocess-call method but would like to get away from the external pdftk dependency so pypdf2 is great:
- text field values show on the generated pdf, but checkbox field values (populated with True or False) don't seem to show up
- there are some vertical shifting issues in pypdf2 output as compared to pdftk output - a few of the fields get bumped up or down on the generated pdf

**Comment by shurshilov** [comment]
[output.pdf](https://github.com/mstamy2/PyPDF2/files/1889844/output.pdf)
Does not work in the fields in this file, for example, the first field for the phone, the second one for some reason works and a few more fields, so the fix is ​​not working

**Comment by saipawan999** [comment]
Hi i am facing the same issue...i have tried setting need lreferences true also.when i edited pdf using pypdf2 some fields are displaying correctly and some are displaying only after i click on that filed.Please help me out on this issue as it is blocking me from the work.
Thank you

**Comment by fvw222** [comment]
The code works great! but only for PDFs with one page. I tried splitting my PDF into several one page files and looped through it. This worked great but when I merged them back together, the click-to-reveal-text problem reemerged. The problem lies in the .addPage command for the PdfFileWritter. 
```
for page_number in range(pdf.total_pages):
    pdf2.addPage(pdf.getPage(page_number))
    pdf2.updatePageFormFieldValues(pdf2.getPage(page_number), field_dictionary)
```
When I enter this and try to save, I get an error message: "TypeError: argument should be integer or None, not 'NullObject'" It seems that the .addpage does not append the filewriter but treats each page as a seperate object. Does some one have a solution for this?

**Comment by aatish29** [comment]
Hi All,

Thanks for your help.

I was able to view the text fields of the PDF Form using pypdf2. But still could not figure out to make the visibility(need appearances) of the checkbox of PDF Form.

Thanks in advance.

**Comment by karnh** [comment]
I found answer for checkboxes issue at https://stackoverflow.com/questions/35538851/how-to-check-uncheck-checkboxes-in-a-pdf-with-python-preferably-pypdf2.

```
def updateCheckboxValues(page, fields):

    for j in range(0, len(page['/Annots'])):
        writer_annot = page['/Annots'][j].getObject()
        for field in fields:
            if writer_annot.get('/T') == field:
                writer_annot.update({
                    NameObject("/V"): NameObject(fields[field]),
                    NameObject("/AS"): NameObject(fields[field])
                })
```

And as the comment says checked value could be anything depending on how the form was created. Its present in '/AP'. Which I extracted using `list(writer_annot.get('/AP').get('/N').keys())[0]`.

**Comment by madornetto** [comment]
ok, I have implemented the above and it works on my pdf forms however once the form has been updated by the python it can't be run through the code a second time, as getFormFields returns an empty list.  If I open the pdf in Adobe and add a space to the end of a form field value and save, run the code on the form again, getFormFields returns the correct list.

**Comment by mdodoo** [comment]
I am having the same problem: fields not visible fixed by above-mentioned set_need_appearances_writer() approach but getFormFields/pdftk dump_data_fields does not see them. 

In addition, it looks like my fonts somehow get messed up: one of the fields is actually a barcode font. But, after going through PyPDF2 to make a copy with updated fields, the field that uses the barcode font in the original copy now uses one of the other fonts.

**Comment by willingham** [comment]
I'm experiencing the same click-to-reveal-text issue.  Here are a few interesting things I have noticed.

- When using some of the irs forms e.g. https://www.irs.gov/pub/irs-pdf/f1095c.pdf, the issue doesn't happen.
- When creating forms PDFElement's 'Form Field Recognition' feature, the issue doesn't happen.  
- When manually adding fields using PDFElement, the issue happens sometimes.

**Comment by mjl** [comment]
> If I open the updated pdf in Adobe and add a space to the end of a form field value and save, run the code on the form again, getFormFields returns the correct list.

For reference, I just stumbled on the same issue. The problem is that the generated pdf does not have an /AcroForm, and the easiest solution is probably to copy it over from the source file like this:

    trailer = reader.trailer["/Root"]["/AcroForm"]
    writer._root_object.update({
            NameObject('/AcroForm'): trailer
        })

**Comment by Nivatius** [comment]
@mjl can you elaborate how to implement those lines?

**Comment by zoiiieee** [comment]
anyone figure out a solution to set /NeedAppearance for a pdf with multiple pages?

**Comment by sstamand** [comment]
To include multiple pages to the output PDF, I added the pages from the template onto the outpuf file....

if "/AcroForm" in pdf2._root_object:
        pdf2._root_object["/AcroForm"].update(
                {NameObject("/NeedAppearances"): BooleanObject(True)})
        pdf2.addPage(pdf.getPage(0))
        pdf2.updatePageFormFieldValues(pdf2.getPage(0), student_data)
        **pdf2.addPage(pdf.getPage(1))
        pdf2.addPage(pdf.getPage(2))**
        outputStream = open(cs_output, "wb")
        pdf2.write(outputStream)
        outputStream.close()

**Comment by zoiiieee** [comment]
> To include multiple pages to the output PDF, I added the pages from the template onto the outpuf file....

I tried the same thing but Need Appearances seems to apply only to the first page. All the fields on the second page are hidden until focused.

**Comment by jeffneuen** [comment]
Does anyone have a working fix for this issue for multi-page PDFs?

**Comment by brunnurs** [comment]
> @mjl can you elaborate how to implement those lines?

You will have a _pdf-reader_ reading in the origin file and a _pdf-writer_, creating the new pdf (see code of @Tromar44 above). Now you simply need to "copy" over the AcroForm with the lines @mjl presented.

**Comment by hchillon** [comment]
Joining all these explanations I build this code, following **brunnurs** previous comment It works perfectly for me:

- Fill forms with mote than one page
- Work well with checkboxes
- You can see all filled data with any pdf reader


from PyPDF2 import PdfFileReader, PdfFileWriter
from PyPDF2.generic import BooleanObject, NameObject, IndirectObject, TextStringObject


def set_need_appearances_writer(writer):
    # See 12.7.2 and 7.7.2 for more information:
    # http://www.adobe.com/content/dam/acom/en/devnet/acrobat/pdfs/PDF32000_2008.pdf
    try:
        catalog = writer._root_object
        # get the AcroForm tree and add "/NeedAppearances attribute
        if "/AcroForm" not in catalog:
            writer._root_object.update({
                NameObject("/AcroForm"): IndirectObject(len(writer._objects), 0, writer)})

        need_appearances = NameObject("/NeedAppearances")
        writer._root_object["/AcroForm"][need_appearances] = BooleanObject(True)
        return writer

    except Exception as e:
        print('set_need_appearances_writer() catch : ', repr(e))
        return writer



class PdfFileFiller(object):
    
    def __init__(self, infile):
        
        self.pdf = PdfFileReader(open(infile, "rb"), strict=False)
        if "/AcroForm" in self.pdf.trailer["/Root"]:
            self.pdf.trailer["/Root"]["/AcroForm"].update(
            {NameObject("/NeedAppearances"): BooleanObject(True)})
        
   
        
    def update_form_values(self, outfile, newvals=None, newchecks=None):

        self.pdf2 = MyPdfFileWriter()

        # Este fragmento habilita que los campos se ven en cualquier reader
        trailer = self.pdf.trailer["/Root"]["/AcroForm"]
        self.pdf2._root_object.update({
            NameObject('/AcroForm'): trailer})
        # ..............................................................


        set_need_appearances_writer(self.pdf2)
        if "/AcroForm" in self.pdf2._root_object:
            self.pdf2._root_object["/AcroForm"].update(
            {NameObject("/NeedAppearances"): BooleanObject(True)})
        
        for i in range(self.pdf.getNumPages()):
            self.pdf2.addPage(self.pdf.getPage(i))
            self.pdf2.updatePageFormFieldValues(self.pdf2.getPage(i), newvals)
            self.pdf2.updatePageFormCheckboxValues(self.pdf2.getPage(i), newchecks)

        with open(outfile, 'wb') as out:
            self.pdf2.write(out)
        
            
class MyPdfFileWriter(PdfFileWriter):
    
    def __init__(self):
        super().__init__()
        
    def updatePageFormCheckboxValues(self, page, fields):
        '''
        Añadido por Hugo y obtenido en
        https://stackoverflow.com/questions/35538851/how-to-check-uncheck-checkboxes-in-a-pdf-with-python-preferably-pypdf2
        '''

        for j in range(0, len(page['/Annots'])):
            writer_annot = page['/Annots'][j].getObject()
            for field in fields:
                if writer_annot.get('/T') == field:
                    #print('-------------------------------------')
                    #print('     FOUND', field)
                    #print(writer_annot.get('/V'))
                    writer_annot.update({
                        NameObject("/V"): NameObject(fields[field]),
                        NameObject("/AS"): NameObject(fields[field])
                    })


if __name__ == '__main__':

        origin = '900in.pdf'
        destination = '900out.pdf'
        newvals = {"IDETNCON[0]": "A123456T",
                    "NOMSOL[0]": "ARTICA S.L."}
        newchecks={"periodeliq1[0]": "/1"}
                    

        c = PdfFileFiller(origin)
        c. update_form_values(outfile=destination,
                              newvals=newvals,
                              newchecks=newchecks)

**Comment by hchillon** [comment]
From all those explanations I arrived (as brunnurs stated) to this code. It works for me. Fill textentries and checkboxes for multipage pdf and all changes can be seen using any simple pdf reader.

`from PyPDF2 import PdfFileReader, PdfFileWriter
from PyPDF2.generic import BooleanObject, NameObject, IndirectObject, TextStringObject


def set_need_appearances_writer(writer):

    try:
        catalog = writer._root_object
        # get the AcroForm tree and add "/NeedAppearances attribute
        if "/AcroForm" not in catalog:
            writer._root_object.update({
                NameObject("/AcroForm"): IndirectObject(len(writer._objects), 0, writer)})

        need_appearances = NameObject("/NeedAppearances")
        writer._root_object["/AcroForm"][need_appearances] = BooleanObject(True)
        return writer

    except Exception as e:
        print('set_need_appearances_writer() catch : ', repr(e))
        return writer



class PdfFileFiller(object):
    
    def __init__(self, infile):
        
        self.pdf = PdfFileReader(open(infile, "rb"), strict=False)
        if "/AcroForm" in self.pdf.trailer["/Root"]:
            self.pdf.trailer["/Root"]["/AcroForm"].update(
            {NameObject("/NeedAppearances"): BooleanObject(True)})
        
   
        
    def update_form_values(self, outfile, newvals=None, newchecks=None):

        self.pdf2 = MyPdfFileWriter()


        trailer = self.pdf.trailer["/Root"]["/AcroForm"]
        self.pdf2._root_object.update({
            NameObject('/AcroForm'): trailer})

        set_need_appearances_writer(self.pdf2)
        if "/AcroForm" in self.pdf2._root_object:
            self.pdf2._root_object["/AcroForm"].update(
            {NameObject("/NeedAppearances"): BooleanObject(True)})
        
        for i in range(self.pdf.getNumPages()):
            self.pdf2.addPage(self.pdf.getPage(i))
            self.pdf2.updatePageFormFieldValues(self.pdf2.getPage(i), newvals)
            self.pdf2.updatePageFormCheckboxValues(self.pdf2.getPage(i), newchecks)

        with open(outfile, 'wb') as out:
            self.pdf2.write(out)
        
            
class MyPdfFileWriter(PdfFileWriter):
    
    def __init__(self):
        super().__init__()
        
    def updatePageFormCheckboxValues(self, page, fields):

        for j in range(0, len(page['/Annots'])):
            writer_annot = page['/Annots'][j].getObject()
            for field in fields:
                if writer_annot.get('/T') == field:
                    #print('-------------------------------------')
                    #print('     FOUND', field)
                    #print(writer_annot.get('/V'))
                    writer_annot.update({
                        NameObject("/V"): NameObject(fields[field]),
                        NameObject("/AS"): NameObject(fields[field])
                    })


if __name__ == '__main__':

        origin = '900in.pdf'
        destination = '900out.pdf'
        newvals = {"IDETNCON[0]": "A123456T",
                    "NOMSOL[0]": "ARTICA S.L."}
        newchecks={"periodeliq1[0]": "/1"}
                    

        c = PdfFileFiller(origin)
        c. update_form_values(outfile=destination,
                              newvals=newvals,
                              newchecks=newchecks)`

**Comment by hchillon** [comment]
Sorry. Last code fails for checkboxes for some pdf readers. So I made a modification in class MyPdfWriter:

`def updatePageFormCheckboxValues(self, page, fields):
 
        for j in range(0, len(page['/Annots'])):
            writer_annot = page['/Annots'][j].getObject()
            for field in fields:
                if writer_annot.get('/T') == field:
                    print('-------------------------------------')
                    print('     FOUND', field)
                    print(writer_annot.get('/V'))
                    if fields[field] in ('/1', '/Yes'):
                        writer_annot.update({
                            NameObject("/V"): NameObject(fields[field]),
                            NameObject("/AS"): NameObject(fields[field])
                        })
                        print(fields[field], ### writer_annot.get('/V'))`

**Comment by hchillon** [comment]
Last code fails for checkboxes using some pdf readers. I modified my MyPdfWriter class:

`def updatePageFormCheckboxValues(self, page, fields):
 
        for j in range(0, len(page['/Annots'])):
            writer_annot = page['/Annots'][j].getObject()
            for field in fields:
                if writer_annot.get('/T') == field:
                    if fields[field] in ('/1', '/Yes'): # You choose which values use in your code
                        writer_annot.update({
                            NameObject("/V"): NameObject(fields[field]),
                            NameObject("/AS"): NameObject(fields[field])
                        })`

**Comment by giorgio-pap** [comment]
I am still having issues in showing filled boxed outside of Adobe Acrobat. 
```
from PyPDF2 import PdfFileWriter, PdfFileReader
from PyPDF2.generic import BooleanObject, NameObject, IndirectObject

def set_need_appearances_writer(writer: PdfFileWriter):
    # See 12.7.2 and 7.7.2 for more information: http://www.adobe.com/content/dam/acom/en/devnet/acrobat/pdfs/PDF32000_2008.pdf
    try:
        catalog = writer._root_object
        # get the AcroForm tree
        if "/AcroForm" not in catalog:
            writer._root_object.update({
                NameObject("/AcroForm"): IndirectObject(len(writer._objects), 0, writer)})

        need_appearances = NameObject("/NeedAppearances")
        writer._root_object["/AcroForm"][need_appearances] = BooleanObject(True)
        return writer

    except Exception as e:
        print('set_need_appearances_writer() catch : ', repr(e))
        return writer

infile = "input.pdf"
outfile = "output.pdf"

pdf = PdfFileReader(open(infile, "rb"), strict=False)
if "/AcroForm" in pdf.trailer["/Root"]:
    pdf.trailer["/Root"]["/AcroForm"].update(
        {NameObject("/NeedAppearances"): BooleanObject(True)})

pdf2 = PdfFileWriter()
set_need_appearances_writer(pdf2)
if "/AcroForm" in pdf2._root_object:
    pdf2._root_object["/AcroForm"].update(
        {NameObject("/NeedAppearances"): BooleanObject(True)})

field_dictionary = {"iban1_part1": "DE", "Model": "Tacoma"}

pdf2.addPage(pdf.getPage(0))
pdf2.updatePageFormFieldValues(pdf2.getPage(0), field_dictionary)

outputStream = open(outfile, "wb")
pdf2.write(outputStream)
```

Some boxes are showing properly, some are not - when outside of Acrobat and I need to click on them to show the content.

I also did the same using pdfrw but I got stuck exactly at the same problem.

**Comment by hchillon** [comment]
Hi, giorgio-pap. I'm using the code in a project that I'm developing in order  to fill tax forms in Andorra. Because of your comment I have been testing the code and these are my results:

- A lot of problems with Adobe Acrobat 9.0 (Last available version for Manjaro Linux)
- Good results with MasterPDF (https://code-industry.net/masterpdfeditor/)
- Good results with qdpview (https://github.com/bendikro/qpdfview)

As I'm not a Windows user, I don't use Adobe PDF tools. MasterPDF and qpdfview are my best alternatives working with Linux. Can you test your code with these alternatives?

**Comment by hchillon** [comment]
Hi again, giorgio-pap.  Have you check issue #545?

**Comment by giorgio-pap** [comment]
> Hi, giorgio-pap. I'm using the code in a project that I'm developing in order to fill tax forms in Andorra. Because of your comment I have been testing the code and these are my results:
> 
>     * A lot of problems with Adobe Acrobat 9.0 (Last available version for Manjaro Linux)
> 
>     * Good results with MasterPDF (https://code-industry.net/masterpdfeditor/)
> 
>     * Good results with qdpview (https://github.com/bendikro/qpdfview)
> 
> 
> As I'm not a Windows user, I don't use Adobe PDF tools. MasterPDF and qpdfview are my best alternatives working with Linux. Can you test your code with these alternatives?

Thanks a lot for your reply! Unfortunately, this script is meant to work for a whole company. So it is necesarry that the ouptut is steady with every most common reading softwares, since I can not require anyone to install anything.

**Comment by SlawoKleeb** [comment]
> @ademidun That worked perfectly (I'd high five you right now if I could)! Thank you very much! For anyone else interested, the following worked for me:
> 
> ```
> from PyPDF2 import PdfFileWriter, PdfFileReader
> from PyPDF2.generic import BooleanObject, NameObject, IndirectObject
> 
> def set_need_appearances_writer(writer: PdfFileWriter):
>     # See 12.7.2 and 7.7.2 for more information: http://www.adobe.com/content/dam/acom/en/devnet/acrobat/pdfs/PDF32000_2008.pdf
>     try:
>         catalog = writer._root_object
>         # get the AcroForm tree
>         if "/AcroForm" not in catalog:
>             writer._root_object.update({
>                 NameObject("/AcroForm"): IndirectObject(len(writer._objects), 0, writer)})
> 
>         need_appearances = NameObject("/NeedAppearances")
>         writer._root_object["/AcroForm"][need_appearances] = BooleanObject(True)
>         return writer
> 
>     except Exception as e:
>         print('set_need_appearances_writer() catch : ', repr(e))
>         return writer
> 
> infile = "myInputPdf.pdf"
> outfile = "myOutputPdf.pdf"
> 
> pdf = PdfFileReader(open(infile, "rb"), strict=False)
> if "/AcroForm" in pdf.trailer["/Root"]:
>     pdf.trailer["/Root"]["/AcroForm"].update(
>         {NameObject("/NeedAppearances"): BooleanObject(True)})
> 
> pdf2 = PdfFileWriter()
> set_need_appearances_writer(pdf2)
> if "/AcroForm" in pdf2._root_object:
>     pdf2._root_object["/AcroForm"].update(
>         {NameObject("/NeedAppearances"): BooleanObject(True)})
> 
> field_dictionary = {"Make": "Toyota", "Model": "Tacoma"}
> 
> pdf2.addPage(pdf.getPage(0))
> pdf2.updatePageFormFieldValues(pdf2.getPage(0), field_dictionary)
> 
> outputStream = open(outfile, "wb")
> pdf2.write(outputStream)
> ```

purrs like a kitten :-)

**Comment by VoidIsEverywhere** [comment]
> I am still having issues in showing filled boxed outside of Adobe Acrobat.
> 
> ```
> from PyPDF2 import PdfFileWriter, PdfFileReader
> from PyPDF2.generic import BooleanObject, NameObject, IndirectObject
> 
> def set_need_appearances_writer(writer: PdfFileWriter):
>     # See 12.7.2 and 7.7.2 for more information: http://www.adobe.com/content/dam/acom/en/devnet/acrobat/pdfs/PDF32000_2008.pdf
>     try:
>         catalog = writer._root_object
>         # get the AcroForm tree
>         if "/AcroForm" not in catalog:
>             writer._root_object.update({
>                 NameObject("/AcroForm"): IndirectObject(len(writer._objects), 0, writer)})
> 
>         need_appearances = NameObject("/NeedAppearances")
>         writer._root_object["/AcroForm"][need_appearances] = BooleanObject(True)
>         return writer
> 
>     except Exception as e:
>         print('set_need_appearances_writer() catch : ', repr(e))
>         return writer
> 
> infile = "input.pdf"
> outfile = "output.pdf"
> 
> pdf = PdfFileReader(open(infile, "rb"), strict=False)
> if "/AcroForm" in pdf.trailer["/Root"]:
>     pdf.trailer["/Root"]["/AcroForm"].update(
>         {NameObject("/NeedAppearances"): BooleanObject(True)})
> 
> pdf2 = PdfFileWriter()
> set_need_appearances_writer(pdf2)
> if "/AcroForm" in pdf2._root_object:
>     pdf2._root_object["/AcroForm"].update(
>         {NameObject("/NeedAppearances"): BooleanObject(True)})
> 
> field_dictionary = {"iban1_part1": "DE", "Model": "Tacoma"}
> 
> pdf2.addPage(pdf.getPage(0))
> pdf2.updatePageFormFieldValues(pdf2.getPage(0), field_dictionary)
> 
> outputStream = open(outfile, "wb")
> pdf2.write(outputStream)
> ```
> 
> Some boxes are showing properly, some are not - when outside of Acrobat and I need to click on them to show the content.
> 
> I also did the same using pdfrw but I got stuck exactly at the same problem.
I tried this this code and nothing appears on linux defualt pdfveiwer but will on adobe and if you open it on gmail on most platforms.  But, oniphones it only shows some fields I poked around a bit and I think might have something to do with the PDF format but could not solve it with this python tool.
I found another that did the job and was viewable from all platforms tested.
https://www.blog.pythonlibrary.org/2018/05/22/filling-pdf-forms-with-python/
The code I used is at the very bottom titled "using the pdfforms package".
The down side is the code so far hasn't been successfully ran on anything but, Linux
and It doesn't click boxes.

**Comment by EricSamsonCarto** [comment]
Might be a separate issue, but I am having a similar problem with PdfFileMerger(). After merging two PDFs together, one having filled forms, the filled form values do not carry over to the final merged version. However, the values do appear when clicking into one of the forms, weirdly enough. I was wondering if I could apply the above logic, but for PdfFileMerger() instead of PdfFileWriter(), but I'm not sure how to implement that. The append section of my code, simplified:

temp_pdf = r"path.pdf"
appendpdf = r"path.pdf"
merger = PdfFileMerger()
merger.append(PdfFileReader(temp_pdf))
merger.append(PdfFileReader(appendpdf))
merger.write(temp_pdf)
merger.close()

The temp_pdf is the one with forms, the appendpdf is typically an image. I'm writing the final merged PDF back to the temp_pdf to overwrite it, that might be a problem, im not sure.

**Comment by charalamm** [comment]
Hello everyone! I tried @hchillon andit works fine for me. Thanks @hchillon. Thanks you for sharing it!! 
I would like to note that the code does not to the job when the `newvals` dict has empty values. For example `newvals =  {'something':'', 'smth2':'smth'}` would make again the values appear only when the field is clicked. I am posting this for everyone who has a hard time figuring out why it doesn;t work.

> 
> 
> From all those explanations I arrived (as brunnurs stated) to this code. It works for me. Fill textentries and checkboxes for multipage pdf and all changes can be seen using any simple pdf reader.
> 
> `from PyPDF2 import PdfFileReader, PdfFileWriter
> from PyPDF2.generic import BooleanObject, NameObject, IndirectObject, TextStringObject
> 
> def set_need_appearances_writer(writer):
> 
> ```
> try:
>     catalog = writer._root_object
>     # get the AcroForm tree and add "/NeedAppearances attribute
>     if "/AcroForm" not in catalog:
>         writer._root_object.update({
>             NameObject("/AcroForm"): IndirectObject(len(writer._objects), 0, writer)})
> 
>     need_appearances = NameObject("/NeedAppearances")
>     writer._root_object["/AcroForm"][need_appearances] = BooleanObject(True)
>     return writer
> 
> except Exception as e:
>     print('set_need_appearances_writer() catch : ', repr(e))
>     return writer
> ```
> 
> class PdfFileFiller(object):
> 
> ```
> def __init__(self, infile):
>     
>     self.pdf = PdfFileReader(open(infile, "rb"), strict=False)
>     if "/AcroForm" in self.pdf.trailer["/Root"]:
>         self.pdf.trailer["/Root"]["/AcroForm"].update(
>         {NameObject("/NeedAppearances"): BooleanObject(True)})
>     
> 
>     
> def update_form_values(self, outfile, newvals=None, newchecks=None):
> 
>     self.pdf2 = MyPdfFileWriter()
> 
> 
>     trailer = self.pdf.trailer["/Root"]["/AcroForm"]
>     self.pdf2._root_object.update({
>         NameObject('/AcroForm'): trailer})
> 
>     set_need_appearances_writer(self.pdf2)
>     if "/AcroForm" in self.pdf2._root_object:
>         self.pdf2._root_object["/AcroForm"].update(
>         {NameObject("/NeedAppearances"): BooleanObject(True)})
>     
>     for i in range(self.pdf.getNumPages()):
>         self.pdf2.addPage(self.pdf.getPage(i))
>         self.pdf2.updatePageFormFieldValues(self.pdf2.getPage(i), newvals)
>         self.pdf2.updatePageFormCheckboxValues(self.pdf2.getPage(i), newchecks)
> 
>     with open(outfile, 'wb') as out:
>         self.pdf2.write(out)
> ```
> 
> class MyPdfFileWriter(PdfFileWriter):
> 
> ```
> def __init__(self):
>     super().__init__()
>     
> def updatePageFormCheckboxValues(self, page, fields):
> 
>     for j in range(0, len(page['/Annots'])):
>         writer_annot = page['/Annots'][j].getObject()
>         for field in fields:
>             if writer_annot.get('/T') == field:
>                 #print('-------------------------------------')
>                 #print('     FOUND', field)
>                 #print(writer_annot.get('/V'))
>                 writer_annot.update({
>                     NameObject("/V"): NameObject(fields[field]),
>                     NameObject("/AS"): NameObject(fields[field])
>                 })
> ```
> 
> if **name** == '**main**':
> 
> ```
>     origin = '900in.pdf'
>     destination = '900out.pdf'
>     newvals = {"IDETNCON[0]": "A123456T",
>                 "NOMSOL[0]": "ARTICA S.L."}
>     newchecks={"periodeliq1[0]": "/1"}
>                 
> 
>     c = PdfFileFiller(origin)
>     c. update_form_values(outfile=destination,
>                           newvals=newvals,
>                           newchecks=newchecks)`
> ```

**Comment by njk639** [comment]
@charalamm  Awesome work. Mind if I adapt it for a project I'm using? I'll be modifying a bit for my needs to avoid hard coding the newvals and newchecks dicts.

**Comment by CTMBNara** [comment]
If you suddenly help someone.
I had the same issue, [solution](https://github.com/mstamy2/PyPDF2/issues/355#issuecomment-615478295) didn't help for PDF Reader Pro and for standard preview function on Mac OS.
Comparing several pdf files, the following helped me:
```
for pageNumber in range(writer.getNumPages()):
    annotationsCount = len(writer.getPage(pageNumber)['/Annots'])
    for annotationNumber in range(annotationsCount):
        annotation = writer.getPage(pageNumber)['/Annots'][annotationNumber].getObject()
        annotation[NameObject('/AP')] = annotation['/AP']['/N']
```

**Comment by ale-rt** [comment]
I think the issue is related to the writer not being initialized properly.
I resolved the issue copying some data from the reader, see:
```
#!/usr/bin/env python3
from PyPDF4.generic import NameObject
from PyPDF4.generic import TextStringObject
from PyPDF4.pdf import PdfFileReader
from PyPDF4.pdf import PdfFileWriter

import random
import sys

reader = PdfFileReader(sys.argv[1])

writer = PdfFileWriter()
# Try to "clone" the original one (note the library has cloneDocumentFromReader)
# but the render pdf is blank
writer.appendPagesFromReader(reader)
writer._info = reader.trailer["/Info"]
reader_trailer = reader.trailer["/Root"]
writer._root_object.update(
    {
        key: reader_trailer[key]
        for key in reader_trailer
        if key in ("/AcroForm", "/Lang", "/MarkInfo")
    }
)

page = writer.getPage(0)

params = {"Foo": "Bar"}

# Inspired by updatePageFormFieldValues but also handle checkboxes
for annot in page["/Annots"]:
    writer_annot = annot.getObject()
    field = writer_annot["/T"]
    if writer_annot["/FT"] == "/Btn":
        value = params.get(field, random.getrandbits(1))
        if value:
            writer_annot.update(
                {
                    NameObject("/AS"): NameObject("/On"),
                    NameObject("/V"): NameObject("/On"),
                }
            )
    elif writer_annot["/FT"] == "/Tx":
        value = params.get(field, field)
        writer_annot.update(
            {
                NameObject("/V"): TextStringObject(value),
            }
        )

with open(sys.argv[2], "wb") as f:
    writer.write(f)
```

See also https://stackoverflow.com/a/66388344/646005

**Comment by Dpats13** [comment]
After reading through this thread and trying many of the suggested solutions above, I still was getting strange behavior when previewing the PDF in an application that was not dedicated to viewing / editing PDFs (ex. mobile email client). The PDF would display without showing any of the filled form fields. After piecing together a few solutions mentioned above, I realized that the order is critical in getting the correct behavior. Here is the solution I am using today:

`

    def _set_need_appearances_writer(writer: PdfFileWriter):
        """
        Enables PDF filled form values to be visible on the final PDF results

        NOTE: See 12.7.2 and 7.7.2 for more information:
        http://www.adobe.com/content/dam/acom/en/devnet/acrobat/pdfs/PDF32000_2008.pdf
        """
        try:
            catalog = writer._root_object
            # get the AcroForm tree
            if "/AcroForm" not in catalog:
                writer._root_object.update({
                    NameObject("/AcroForm"): IndirectObject(len(writer._objects), 0, writer)
                })

            need_appearances = NameObject("/NeedAppearances")
            writer._root_object["/AcroForm"][need_appearances] = BooleanObject(True)
            # del writer._root_object["/AcroForm"]['NeedAppearances']
            return writer

        except Exception as e:
            print('set_need_appearances_writer() catch : ', repr(e))
            return writer

    def _lock_form_fields(cls, page):
        """
        Locks all form fields on the given PyPdf2 Page object
        """
        for j in range(0, len(page['/Annots'])):
            writer_annot = page['/Annots'][j].getObject()
            if writer_annot.get('/T'):
                writer_annot.update({
                    NameObject("/Ff"): NumberObject(1)
                })

# Putting it all togeher
    def _init_pdf_writer_from_reader(cls, reader: PdfFileReader) -> PdfFileWriter:
        """
        Initializes a PdfFileWriter that can be used to write data to the given PDF
        stored inside of the PdfFileReader.

        IMPORTANT: Using this init function ensures that the data written is visible
                   both in a PDF Viewer Application and in a Preview context (i.e. an email client)
        """
        if not reader or reader.getNumPages() == 0:
            raise Exception(f"Error initializing PdfFileWriter, given PdfFileReader "
                            f"is either null or contains no pages.")

        pdf_writer = PdfFileWriter()

        # Add all PDF pages from reader -> writer
        pdf_writer.appendPagesFromReader(reader)

        # Copy over additional data from reader -> writer
        pdf_writer._info = reader.trailer["/Info"]
        reader_trailer = reader.trailer["/Root"]
        pdf_writer._root_object.update(
            {
                key: reader_trailer[key]
                for key in reader_trailer
                if key in ("/AcroForm", "/Lang", "/MarkInfo")
            }
        )

        # Set written data appearances to be visible
        cls._set_need_appearances_writer(pdf_writer)
        if "/AcroForm" in pdf_writer._root_object:
            pdf_writer._root_object["/AcroForm"].update(
                {NameObject("/NeedAppearances"): BooleanObject(True)})

        return pdf_writer
`

By initializing the PDF Writer correctly we ensure that the data written to the PDFs form fields will be visible without having to click the field in a PDF viewer application. We also guarantee it will be visible in a non-pdf viewing specific application which is important if your client / end-user is using an app that you can not be sure of what it will be to view the PDF. Lastly I included a method to lock the fields on a given PDF page, that way it is no longer editable by your end-user (if this is desired behavior).

Thanks to @ale-rt and many others above.

**Comment by apteryxlabs** [comment]
@Dpats13 is your code part of a broader object definition? I'm thrown by the cls args.

**Comment by Dpats13** [comment]
@apteryxlabs ya, you can ignore those.

**Comment by Tromar44** [comment]
@Dpats13 Id like to implement your solution but my python/programming skills are not great. Do you mind posting some working code assuming variables similar to below? 
`
infile = 'myInputPdf.pdf'
outfile = 'myOutputPdf.pdf'
field_dictionary = {'foo':'bar'}
`
One of the old solutions above offered by @ademidun via @Tromar44 above still works well for me for filling pdf forms and reading them but trying to go back and programmatically (ex. PyPDF2, pdfminer) read the content of those filled forms returns empty fields (ie. I can manually open the PDF and see the content of those fields without clicking them but reading them via python returns empty fields). If I manually open the PDF and save it before closing it, then I am able to programmatically read the fields.
Any demo/example of your solution would be greatly appreciated - thanks!

**Comment by MartinThoma** [comment]
I still see this issue in the Atril document viewer for `test_fill_form`.

**Comment by MartinThoma** [comment]
Aparently some people had luck with

```python
# Set /NeedAppearances
writer.set_need_appearances_writer()

# Make it read-only with /Ff:
writer.updatePageFormFieldValues(writer.getPage(0), {"foo": "some filled in text"}, flags=1)
```

However, at least with Evince this doesn't work. And the Google Chrome PDF viewer always shows the filled fields.

**Comment by mjl** [comment]
I see quite a few comments that /NeedAppearances makes it work, but I'm sorry, that is not a general solution. This hints to the reader app that it needs to do some work to render correct form fields, but there are a lot of reader apps out there that do not honor that flag or do it badly.

What one needs to do is to really go over all the form fields and render them (ie. text input field -> add an Appearance Stream /AP that renders the entered value, checkboxes -> add Appearance State /AS that shows the field checked, other field types probably need even more work, this I have not investigated because I did not need those thus far). 

What I ended up doing is inspect Acrobat generated forms and emulating that. I think I used `qpdf --show-object` to dissect the pdfs.

This comment helped me lots to get me started: https://github.com/pmaupin/pdfrw/issues/84#issuecomment-445303928

**Comment by MartinThoma** [comment]
Summarizing some ideas:

1. We might need to set / adjust some annotations, see https://github.com/py-pdf/PyPDF2/issues/546#issuecomment-1179322743
2. We might need to set / write the `Fields` dictionary
3. The `/NeedAppearances` seems not to help as it was added via `writer.set_need_appearances_writer()`

**Comment by fidoriel** [comment]
Yesterday I took a deep dive into the PDF standard. I am 99% confident that this issue originates, like @mjl said, in a missing apperance stream. Today I was able to append a apperance Stream to the form field. The content of the filled field is now visible in Acrobat, SumatraPDF, Okular, Chrome, Firefox, Edge and it does also print.
Now I am experiencing issues the layout of the text and special characters. I hope that I can solve them soon to be able to submit a PR.

**Comment by codigovision** [comment]
Here's what worked for me, based on the PR from @fidoriel, but without compressing the stream, which caused the preview not to work for me. This is a simplified version just to show how updating the stream and associated dictionary works. It's working to show the filled in form fields and set them to read only. I've tested it in Apple preview for OSX and IOS, and also in Acrobat Pro DC. 

```python
    # Field data.
    data = {"field_name": "some value"}

    # Get template.
    template = PdfReader("template.pdf", strict=False)

    # Initialize writer.
    writer = PdfWriter()

    # Add the template page to the writer.
    writer.add_page(template.pages[0])

    # Get page annotations.
    page_annotations = writer.pages[1][PageAttributes.ANNOTS]

    # Loop through page annotations (fields).
    for index in range(len(page_annotations)):  # type: ignore
        # Get annotation object.
        annotation = page_annotations[index].get_object()  # type: ignore

        # Get existing values needed to create the new stream and update the field.
        field = annotation.get(NameObject("/T"))
        new_value = data[field]
        ap = annotation.get(AnnotationDictionaryAttributes.AP)
        x_object = ap.get(NameObject("/N")).get_object()
        font = annotation.get(InteractiveFormDictEntries.DA)
        rect = annotation.get(AnnotationDictionaryAttributes.Rect)

        # Calculate the text position.
        font_size = float(font.split(" ")[1])
        w = round(float(rect[2] - rect[0] - 2), 2)
        h = round(float(rect[3] - rect[1] - 2), 2)
        text_position_h = h / 2 - font_size / 3  # approximation

        # Create a new XObject stream.
        new_stream = f'''
            /Tx BMC 
            q
            1 1 { w } { h } re W n
            BT
            { font }
            2 { text_position_h } Td
            ({ new_value }) Tj
            ET
            Q
            EMC
        '''

        # Update XObject stream.
        x_object._data = encode_pdfdocencoding(new_stream)

        # Update annotation dictionary.
        annotation.update(
            {
                # Update Value.
                NameObject(FieldDictionaryAttributes.V): TextStringObject(
                    new_value
                ),
                # Update Default Value.
                NameObject(FieldDictionaryAttributes.DV): TextStringObject(
                    new_value
                ),
                # Set Read Only flag.
                NameObject(FieldDictionaryAttributes.Ff): NumberObject(
                    FieldFlag(1)
                )
            }
        )

    # write "output".
    with open("output.pdf", "wb") as output_stream:
        writer.write(output_stream)  # type: ignore
```

**Comment by DanielMajer24** [comment]
Hey all, thanks for all the discussion on this topic. This was very helpful to get my script working.  

I got this by creating a user-defined function for updatePageFormFieldValues instead of using the imported one in PyPDF2. This was created based on @ademidun's initial response and later adapted using parts of @ale-rt's answer for PyPDF4 when text isn't showing in some template boxe. I changed the code slightly from to also from PyPDF4 to PyPDF2 for my use case and ensure checkboxes were being selected. Two things to mention about the function: 

> 1) you need to see how the checkboxes have been formatted and whether they use \Yes, \On or \1 when they are selected as they will only engage with their own specific command 

> 2) You need to have the have all the fields of the PDF form present in the field_dictionary when you input values. If you only select the relevant values in the field_dictionary the function will enter the dictionary keys in the text box values if it has not been mentioned. This is most likely an error I have created but with this work simple fix I have been able to get around it... hopefully. 

```python
from PyPDF2 import PdfFileWriter, PdfFileReader
from PyPDF2.generic import BooleanObject, NameObject, IndirectObject, TextStringObject

#inital functions

def set_need_appearances_writer(writer: PdfFileWriter):
    # See 12.7.2 and 7.7.2 for more information: http://www.adobe.com/content/dam/acom/en/devnet/acrobat/pdfs/PDF32000_2008.pdf
    try:
        catalog = writer._root_object
        time.sleep(1)
        # get the AcroForm tree
        if "/AcroForm" not in catalog:
            writer._root_object.update({
                NameObject("/AcroForm"): IndirectObject(len(writer._objects), 0, writer)})

        need_appearances = NameObject("/NeedAppearances")
        writer._root_object["/AcroForm"][need_appearances] = BooleanObject(True)
        return writer

    except Exception as e:
        print('set_need_appearances_writer() catch : ', repr(e))
        return writer 


def updatePageFormFieldValues(page, fields):
        '''
        Update the form field values for a given page from a fields dictionary.

       This was copied from the PyPDF2 library and adapted for my use case.


        Copy field texts and values from fields to page.
        :param page: Page reference from PDF writer where the annotations
            and field data will be updated.
        :param fields: a Python dictionary of field names (/T) and text
            values (/V)
        '''
        # Iterate through pages, update field values
        for j in range(0, len(page['/Annots'])):
            writer_annot = page['/Annots'][j].getObject()
            field = writer_annot.get('/T') 
            if writer_annot.get("/FT") == "/Btn":
                value = fields.get(field, random.getrandbits(1))
                if value:
                    writer_annot.update(
                        {
                            NameObject("/AS"): NameObject("/On"),
                            NameObject("/V"): NameObject("/On"),
                        }
                    )
            elif writer_annot.get("/FT") == "/Tx":
                value = fields.get(field,field)
                writer_annot.update(
                    {
                        NameObject("/V"): TextStringObject(value),
                    }
                )

def generate_PDF_file(outfile, field_dictionary, infile):

    pdf = PdfFileReader(open(infile, "rb"), strict=False)
    if "/AcroForm" in pdf.trailer["/Root"]:
        pdf.trailer["/Root"]["/AcroForm"].update(
            {NameObject("/NeedAppearances"): BooleanObject(True)})

    pdf2 = PdfFileWriter()

    pdf2._info = pdf.trailer["/Info"]
    reader_trailer = pdf.trailer["/Root"]
    pdf2._root_object.update(
        {
            key: reader_trailer[key]
            for key in reader_trailer
            if key in ("/AcroForm", "/Lang", "/MarkInfo")
        }
    )

    set_need_appearances_writer(pdf2)
    time.sleep(1)
    if "/AcroForm" in pdf2._root_object:
        pdf2._root_object["/AcroForm"].update(
            {NameObject("/NeedAppearances"): BooleanObject(True)})


    #currently the pdfs I have been working with only have one page but I will have to change this part to accommodate multiple page pdfs
    pdf2.addPage(pdf.getPage(0))
    #####This is where I added the user defined function instead of using the PyPDF2 one.  
    updatePageFormFieldValues(page = pdf2.getPage(0), fields = field_dictionary) #changed to user_defined_function

    outputStream = open(outfile, "wb")
    pdf2.write(outputStream)
    
    inputStream.close()
    outputStream.close()
```

**Comment by Bchapp1558** [comment]
Not a solution, unfortunately, I am also in need of help. I am pretty new to programming so trying to incorporate some of these solutions has been tricky. Like everyone else, I am trying to make my data visible on my new form without needing to click on the field first. All of the fields are set to be visible.

`from flask import Blueprint, render_template, request
from PyPDF2 import PdfReader, PdfWriter
from flask_cors import CORS
from PyPDF2.generic import BooleanObject, NameObject


views = Blueprint('views', __name__)
CORS(views)

@views.route('/')
def home():
    return render_template("base.html")

@views.route('/submit', methods=['POST'])
def submit():
    data = request.json
    reader = PdfReader("AF594.pdf")
    writer = PdfWriter()
    page = reader.pages[0]
    fields = reader.get_fields()
    writer.add_page(page)
    lastNameFirst = data['sponsorLastName'] + ", " + data['sponsorFirstName']
    fullPhysicalAddress = data['sponsorAddress'] + ", " + data['city'] + ", " + data['state'] + " " + data['zipCode']
    data.update({ "lastNameFirst": lastNameFirst, "fullPhysicalAddress": fullPhysicalAddress})

    # Puts the X in block 7 for what child support is based on 
    if data['childSupportBasedOnInput'] == "basedOn_DivorceDecree":
        data.update({ "basedOn_DivorceDecree": "X"})
    elif data['childSupportBasedOnInput'] == "basedOn_CourtOrder":
        data.update({ "basedOn_CourtOrder": "X"})
    elif data['childSupportBasedOnInput'] == "basedOn_LegalSeparationAgreement":
        data.update({ "basedOn_LegalSeparationAgreement": "X"})
    elif data['childSupportBasedOnInput'] == "basedOn_WrittenagreementWithChildsCustodian":
        data.update({ "basedOn_WrittenagreementWithChildsCustodian": "X"})

    # set the NeedAppearances flag of the AcroForm to True - This was part of how I tried to solve this issue.
    try:
        catalog = writer._root_object
        if "/AcroForm" not in catalog:
            writer._root_object.update({
                NameObject("/AcroForm"): IndirectObject(len(writer._objects), 0, writer)})
        need_appearances = NameObject("/NeedAppearances")
        writer._root_object["/AcroForm"][need_appearances] = BooleanObject(True)
    except Exception as e:
        print('set_need_appearances_writer() catch : ', repr(e))


    writer.update_page_form_field_values(
        writer.pages[0], data
    )

    fileName594 = data['sponsorFirstName'] + "_" + data['sponsorLastName'] +"_filled594.pdf"
    with open(fileName594, "wb") as output_stream:
        writer.write(output_stream)
    return "success"`

Any help would be appreciated!

**Comment by cryzed** [comment]
The issue seems to be that `PdfWriter.set_need_appearances_writer()` (whether called directly or indirectly by `PdftWriter.update_page_form_field_values()`) fails to create the `/Root/AcroForm` object correctly when it doesn't already exist in the `PdfWriter` object.

From what I can tell, a well-formed `/Root/AcroForm`-object should have keys such as: `/DA`, `/DR`, `/Fields` and `/NeedAppearances`, however `PdfWriter.set_need_appearances_writer()` creates an indirect object (reference) pointing to a badly-formed `/Root/AcroForm`-object with keys `stream/data`, `dict/Filter` and `dict/NeedAppearances` instead.

Another issue seems to be that it matters _when_ we create (clone in our case) the `/Root/AcroForm`-object: before or after having updated the form fields. If I call the function outlined below directly after creating the `PdfWriter` and `PdfReader` instances, before having filled the prompt fields, I run into the same issue again: in the output there's an indirect object pointing to an invalid `/Root/AcroForm`-object. I suspect the issue is the code that serializes the indirect objecting pointing to the `/Root/AcroForm` object, since it clearly works if the `AcroForm`-object lives directly inside `/Root` (not pointed to by an indirect object).

Long story short, after a bunch of debugging with `qpdf --json-output` and comparing the pypdf output with `tkpdf`'s output, I came up with this:


```python
import typing as T

import pypdf.generic
import pypdf.constants
import pypdf


def _fix_acroform(writer: pypdf.PdfWriter, reader: pypdf.PdfReader) -> None:
    root_object = T.cast(pypdf.generic.DictionaryObject, reader.trailer[pypdf.constants.TrailerKeys.ROOT])
    acroform_object = T.cast(pypdf.generic.DictionaryObject, root_object[pypdf.constants.CatalogDictionary.ACRO_FORM])
    writer._root_object[pypdf.generic.NameObject(pypdf.constants.CatalogDictionary.ACRO_FORM)] = acroform_object.clone(writer)
    writer.set_need_appearances_writer()

# ... Update forms etc.

_fix_acroform(writer, reader)
with open("out.pdf", "wb") as file:
    writer.write(file)
```

Call this _at the very end_ just before calling `PdfWriter.write()`. My personal use-case is simply filling a PDF with form fields programatically and then saving the output, I was quite disappointed to learn that this was so badly broken after it is advertised as just working in the documentation.

Another few random notes: We call `PdfObject.clone()` on the existing `/Root/AcroForm`-object instead of setting it directly: This guarantees that the `PdfObject.idnums` are translated to the IDs `PdfWriter` assigned to them, resulting in a well-formed `/Root/AcroForm/Fields` listing with correct references to the fields. Ironically `/Root/AcroForm/Fields` seems to be entirely optional, at least for Adobe Acrobat Reader, so we could get away with omitting it -- just if the references inside are wrong it will fail to display the fields correctly.

To summarize: A well-formed `/Root/AcroForm` object is required with the `/Root/AcroForm/NeedAppearances` flag set to true (and I suspect _only_ that) -- it doesn't seem to matter whether this object lives _directly_ inside the `/Root`-object or is just referenced, however pypdf serializes it incorrectly if its the latter. When we set this object early, directly after instantiating `PdfWriter`, before updating the forms, this object will still get proxied behind an indirect object (and serialized incorrectly). This happens sometimes during the updating of them form fields. To avoid this we call `_fix_acroform()` only directly before writing the output file. The now-resulting PDF displays correctly in Okular, Adobe Acrobat Reader and various pdf.js browser-implementations for me.

**Comment by cryzed** [comment]
Nevermind, doesn't quite work yet, seemingly only render the first page correctly in Adobe Acrobat.

**Comment by brzGatsu** [comment]
Looking forward to your findings. We are currently stuck with pdftk due to this bug... would love to switch to PyPDF once the issue has been resolved. Thanks for taking a look at it!

**Comment by cryzed** [comment]
I read through the PDF docs and took a look at `pdftk`'s source code: implementing appearance streams from scratch is possible, but quite tedious (especially if you want to support most common features). I think I'll go with the `pdftk`-route myself and use it until it becomes unsupported. If that ever happens, I'll take another look at it or hope that PDF is a dead format by that time.

However, I'll reopen my pull request -- the bug that prevents the proper creation of `/Root/AcroForm` _does_ exist, and is fixed by my PR. With this at least, it's possible to render the first page correctly in Adobe Reader and all pages in most other PDF readers.

**Comment by brzGatsu** [comment]
If I understand correctly, with your PR we could split our pdf into single pages, fill the forms individually and then merge them again?

**Comment by pubpub-zz** [comment]
@brzGatsu 
I would have add expected 'PdfWriter.append()' to provide some capability to correctly split documents with fields. Can you confirm it ?

**Comment by cryzed** [comment]
@brzGatsu no, that won't work. The issue is that a PDF reader is supposed to render the appearance streams for all annotations if `/Root/AcroForm/NeedAppearances` is set, when the document is opened. This rendering only happens at runtime (when Adobe Reader displays the file) and is not persisted, so you can't just split the pages and merge them later.

**Comment by csears123** [comment]
I am also experiencing this issue from Adobe Reader, where the NeedAppearances flag is only allowing the first page of the PDF to view the text in the fillable field, as @cryzed documented.  On the second page if I click into the field the text becomes visible, only with the cursor focus.
Really hoping there is a solution to set the appearance-stream for every field if that is the best and most reliable method.  I'll try the example above from @codigovision.

I haven't used pdftk but I will also explore that as an alternative.

**Comment by csears123** [comment]
The example below of adding/updating the appearance-stream seemed to work for a 2-page PDF with fillable fields:
https://github.com/py-pdf/pypdf/issues/355#issuecomment-1238541441 
However the issue persists when merging another fillable PDF form into a single PDF output.  The first 2 pages with the original PDF are still working correctly (after updating the appearance-streams), but all the fields on the 3rd page (from a different PDF) do not show the text, it is hidden behind the input until I click into that field (using Adobe Reader).
Doing a little more debugging it seems the writer annotation's on the 3rd page do not have a 'AP' attribute to begin with, and the function below returns 'None' type:
`ap = writer_annot.get(AnnotationDictionaryAttributes.AP)`
Not sure how to add the missing 'AP' appearance-streams, it's seems complicated.
I also ended up testing pdftk and it just worked first try, no workarounds, issues, or bugs that needed addressing.  I'll probably be scrapping pypdf for now, unless this critical issue is resolved.

**Comment by binury** [comment]
Confirming: this is still an issue. Filled annotations do not display as expected in MacOS/Preview/{Mobile,} Safari. They do render in Chrome & Acrobat

**Comment by michael-hoang** [comment]
Has anyone tried doing this for PyPDF2 v3.0.1?

**Comment by pubpub-zz** [comment]
@binury 
if annotation is displayed in Chrome and Acrobat but not in MacOS Preview/Mobile but, the issue is more likely on this latest program. You may have to identify by your own. At lease technically I have no mean to help you.

**Comment by binury** [comment]
> @binury 
> 
> if annotation is displayed in Chrome and Acrobat but not in MacOS Preview/Mobile but, the issue is more likely on this latest program. You may have to identify by your own. At lease technically I have no mean to help you.

FWIW There is a working implementation available currently pdftk, that does not exhibit the same inconsistent appearance between readers.

I think it's a stretch to claim Pypdf fill is working if the PDFs won't be displayed correctly when viewed on an iPhone…
Sure Acrobat is technically the official PDF viewer and the most spec-compliant...
But nobody is going to use this if it means writing off a huge majority of users viewing the documents on their mobile phones. Not to mention MacOS users. 

In any case… no help needed. just wanted to leave a comment earlier to let you guys know that it's broken still.

**Comment by alenards** [comment]
@binury - I'm having trouble filling only a library named `pdf-annot` on npm. 

Is it this: https://www.npmjs.com/package/ts-pdf-annot

**Comment by pubpub-zz** [comment]
> @binury - I'm having trouble filling only a library named `pdf-annot` on npm. 
> 
> Is it this: https://www.npmjs.com/package/ts-pdf-annot

Your library seems to be JavaScript.  I do not think there is a link with pypdf (python)

**Comment by pubpub-zz** [comment]
> > @binury 
> > 
> > if annotation is displayed in Chrome and Acrobat but not in MacOS Preview/Mobile but, the issue is more likely on this latest program. You may have to identify by your own. At lease technically I have no mean to help you.
> 
> FWIW There is a working implementation available currently (in pdftk, of course, as mentioned in preview comment) that does not exhibit the same inconsistent appearance between readers.
> 
> I think it's a stretch to claim Pypdf fill is working if the PDFs won't be displayed correctly when viewed on an iPhone…
> Sure Acrobat is technically the official PDF viewer and the most spec-compliant...
> But nobody is going to use this if it means writing off a huge majority of users viewing the documents on their mobile phones. Not to mention MacOS users. 
> 
> In any case… no help needed. just wanted to leave a comment earlier to let you guys know that it's broken still.
> 
> In lieu of having access to iOS for testing… There is also a working implementation of creating default appearances in a Node lib called pdf-annot. As a reference that may shed some light on why the default appearances in pypdf aren't working.

@binury 
A PR is under submission to improve field rendering if you want to have a try

**Comment by alenards** [comment]
@pubpub-zz - I think @binury was offering that the referenced library's filled in textfields shown as expected without having to do these workarounds with `/Root/AcroForm`, adjusting the annotations, or putting in the `/NeedsAppearances`. If the handling in that library is correct, it might help resolve the need for workarounds on the `PdfWriter` or these other approaches.

I'm a bit frustrated that the readthedocs for this library make it look like "filling out forms" work [here](https://pypdf.readthedocs.io/en/latest/user/forms.html#filling-out-forms) without these workarounds noted. I was incredibly pumped up to see that the latest release for `pypdf` was on June 4. My use case is filling out forms and having them reliably render in PDF tools (macOS Preview being one of them).

**Comment by pubpub-zz** [comment]
> @pubpub-zz - I think @binury was offering that the referenced library's filled in textfields shown as expected without having to do these workarounds with `/Root/AcroForm`, adjusting the annotations, or putting in the `/NeedsAppearances`. If the handling in that library is correct, it might help resolve the need for workarounds on the `PdfWriter` or these other approaches.
> 
> I'm a bit frustrated that the readthedocs for this library make it look like "filling out forms" work [here](https://pypdf.readthedocs.io/en/latest/user/forms.html#filling-out-forms) without these workarounds noted. I was incredibly pumped up to see that the latest release for `pypdf` was on June 4. My use case is filling out forms and having them reliably render in PDF tools (macOS Preview being one of them). 
> 
> My testing in `ipython` and macOS Preview led to reported behavior (the value were there, but only when I clicked into the fields, and then they will disappear after focus is lost). I'll try calling PdfWriter.set_need_appearances_writer() directly. But I'm probably going to have to look for another library in another language.

Thanks for the comment : I may have read too quickly the message
I understand your position and agree with it. The workaround of needappearance is not the. Best. As said aboveI've produced PR #1864 that is generating the display. It is a first release but If you can test it, it would be great

**Comment by alenards** [comment]
@pubpub-zz - I definitely appreciate the effort for all the folx keeping `pypdf` maintained. All of the `PyPDF2`, `PyPDF4`, all that is dizzying; so relieved to see this library active. 

I'll see if I can look at #1864 - and I'll comment there on that PR thread. 

Thanks again.

**Comment by thomasweiland93** [comment]
Hello, together I have taken a look on #1864 and tested with a PDF from my company. But unfortunately the appearance doesn't look correct on Iphones etc.

The Problem might be following structure on my pdf:

**Parent (writer_parent_annot)**
`{'/DA': '/MyriadPro-Regular 9 Tf 0 0.290 0.439 rg', '/FT': '/Tx', '/Kids': [IndirectObject(88, 0, 2973937844032), IndirectObject(85, 0, 2973937844032)], '/T': '08-Mail2', '/V': '[Nirwana@test.de]'}
`
**Child (writer_annot)**
`{'/AP': {'/N': IndirectObject(89, 0, 2973937844032)}, '/F': 4, '/MK': {}, '/P': IndirectObject(49, 0, 2973937844032), '/Parent': IndirectObject(87, 0, 2973937844032), '/Rect': [246.47300000000001, 232.13200000000001, 513.09299999999996, 220.27699999999999], '/Subtype': '/Widget', '/Type': '/Annot'}`

In this case the code runs just in the else case of update_page_form_field_values and sets AA.AS to /Off

To get a correct view on the Iphone Viewer i have done some small changes in the _writer.py... (but just a messy fix for my current pdf)

I have used the /DA /FT and /V from the writer_parent_annot an the rest from the writer_annot. 

Is this a know Issue?

**Comment by Chrisd204** [comment]
> @ademidun That worked perfectly (I'd high five you right now if I could)! Thank you very much! For anyone else interested, the following worked for me:
> 
> ```python
> from PyPDF2 import PdfFileWriter, PdfFileReader
> from PyPDF2.generic import BooleanObject, NameObject, IndirectObject
> 
> def set_need_appearances_writer(writer: PdfFileWriter):
>     # See 12.7.2 and 7.7.2 for more information: http://www.adobe.com/content/dam/acom/en/devnet/acrobat/pdfs/PDF32000_2008.pdf
>     try:
>         catalog = writer._root_object
>         # get the AcroForm tree
>         if "/AcroForm" not in catalog:
>             writer._root_object.update({
>                 NameObject("/AcroForm"): IndirectObject(len(writer._objects), 0, writer)})
> 
>         need_appearances = NameObject("/NeedAppearances")
>         writer._root_object["/AcroForm"][need_appearances] = BooleanObject(True)
>         return writer
> 
>     except Exception as e:
>         print('set_need_appearances_writer() catch : ', repr(e))
>         return writer
> 
> infile = "input.pdf"
> outfile = "output.pdf"
> 
> reader = PdfFileReader(open(infile, "rb"), strict=False)
> if "/AcroForm" in reader.trailer["/Root"]:
>     reader.trailer["/Root"]["/AcroForm"].update(
>         {NameObject("/NeedAppearances"): BooleanObject(True)})
> 
> writer = PdfFileWriter()
> set_need_appearances_writer(writer)
> if "/AcroForm" in writer._root_object:
>     writer._root_object["/AcroForm"].update(
>         {NameObject("/NeedAppearances"): BooleanObject(True)})
> 
> field_dictionary = {"Make": "Toyota", "Model": "Tacoma"}
> 
> writer.addPage(reader.getPage(0))
> writer.updatePageFormFieldValues(writer.getPage(0), field_dictionary)
> 
> with open(outfile, "wb") as fp:
>     writer.write(fp)
> ```

This solution works!

**Comment by RomHartmann** [comment]
This thread is crazy long, with a lot of old versions and red herrings. 

As of now, this works for me for `pypdf==3.17.4`

```
# ... load file
reader = pypdf.PdfReader(file)
writer = pypdf.PdfWriter()

writer.set_need_appearances_writer()

for page_nr, page in enumerate(reader.pages):
    form_fields = page.get('/Annots')
    if form_fields:
        for field in form_fields.get_object():
            field_object = field.get_object()

            # any other logic
            field_object.update({
                pypdf_generic.NameObject('/V'): pypdf_generic.create_string_object(field_value)
            })
    writer.add_page(page)

# create your output file or stream
writer.write(output_file)
```

Conditions of my test:
- single page PDF
- Only text fields

**Comment by caver456** [comment]
Thanks @RomHartmann that definitely got closer.

In the end, as someone else pointed out, flattening is only part of the answer, and relying on NeedAppearences didn't quite do the trick, so modifying the stream directly gave much better results.  Here's a stackoverflow question spelling out these specific symptoms (not sure if they are the exact same symptoms as everyone else has been experiencing):

[Flattened filled PDF form is 'of invalid format' on Android, and shows blank fields in Chrome extension](https://stackoverflow.com/questions/77857175/flattened-filled-pdf-form-is-of-invalid-format-on-android-and-shows-blank-fie)

and the solution (for our use case, at least) that basically references another solution at https://stackoverflow.com/a/73655665/3577105 - thanks to @JeremyM4n for sure.

**Comment by WMiller256** [comment]
> @ademidun That worked perfectly (I'd high five you right now if I could)! Thank you very much! For anyone else interested, the following worked for me:
> 
> ```python
> from PyPDF2 import PdfFileWriter, PdfFileReader
> from PyPDF2.generic import BooleanObject, NameObject, IndirectObject
> 
> def set_need_appearances_writer(writer: PdfFileWriter):
>     # See 12.7.2 and 7.7.2 for more information: http://www.adobe.com/content/dam/acom/en/devnet/acrobat/pdfs/PDF32000_2008.pdf
>     try:
>         catalog = writer._root_object
>         # get the AcroForm tree
>         if "/AcroForm" not in catalog:
>             writer._root_object.update({
>                 NameObject("/AcroForm"): IndirectObject(len(writer._objects), 0, writer)})
> 
>         need_appearances = NameObject("/NeedAppearances")
>         writer._root_object["/AcroForm"][need_appearances] = BooleanObject(True)
>         return writer
> 
>     except Exception as e:
>         print('set_need_appearances_writer() catch : ', repr(e))
>         return writer
> 
> infile = "input.pdf"
> outfile = "output.pdf"
> 
> reader = PdfFileReader(open(infile, "rb"), strict=False)
> if "/AcroForm" in reader.trailer["/Root"]:
>     reader.trailer["/Root"]["/AcroForm"].update(
>         {NameObject("/NeedAppearances"): BooleanObject(True)})
> 
> writer = PdfFileWriter()
> set_need_appearances_writer(writer)
> if "/AcroForm" in writer._root_object:
>     writer._root_object["/AcroForm"].update(
>         {NameObject("/NeedAppearances"): BooleanObject(True)})
> 
> field_dictionary = {"Make": "Toyota", "Model": "Tacoma"}
> 
> writer.addPage(reader.getPage(0))
> writer.updatePageFormFieldValues(writer.getPage(0), field_dictionary)
> 
> with open(outfile, "wb") as fp:
>     writer.write(fp)
> ```

For future users: this may corrupt the PDF file, if that is the case for you one possible solution is to move the lines 

```
set_need_appearances_writer(writer)
if "/AcroForm" in writer._root_object:
    writer._root_object["/AcroForm"].update(
        {NameObject("/NeedAppearances"): BooleanObject(True)})
```

**After** the page-adding is complete, i.e.

```
from PyPDF2 import PdfFileWriter, PdfFileReader
from PyPDF2.generic import BooleanObject, NameObject, IndirectObject

def set_need_appearances_writer(writer: PdfFileWriter):
    # See 12.7.2 and 7.7.2 for more information: http://www.adobe.com/content/dam/acom/en/devnet/acrobat/pdfs/PDF32000_2008.pdf
    try:
        catalog = writer._root_object
        # get the AcroForm tree
        if "/AcroForm" not in catalog:
            writer._root_object.update({
                NameObject("/AcroForm"): IndirectObject(len(writer._objects), 0, writer)})

        need_appearances = NameObject("/NeedAppearances")
        writer._root_object["/AcroForm"][need_appearances] = BooleanObject(True)
        return writer

    except Exception as e:
        print('set_need_appearances_writer() catch : ', repr(e))
        return writer

infile = "input.pdf"
outfile = "output.pdf"

reader = PdfFileReader(open(infile, "rb"), strict=False)
if "/AcroForm" in reader.trailer["/Root"]:
    reader.trailer["/Root"]["/AcroForm"].update(
        {NameObject("/NeedAppearances"): BooleanObject(True)})

writer = PdfFileWriter()
field_dictionary = {"Make": "Toyota", "Model": "Tacoma"}
writer.addPage(reader.getPage(0))
writer.updatePageFormFieldValues(writer.getPage(0), field_dictionary)

set_need_appearances_writer(writer)
if "/AcroForm" in writer._root_object:
    writer._root_object["/AcroForm"].update(
        {NameObject("/NeedAppearances"): BooleanObject(True)})

with open(outfile, "wb") as fp:
    writer.write(fp)
```

**Comment by stefan6419846** [comment]
While this surely is an old issue, I recommend to switch to the maintained *pypdf* instead which might already solve this out of the box.
